### ISSUE TITLE
validateSync causes error when editing full embedded document

### ISSUE THREAD

**Issue by tomgrossman** [Automattic/mongoose#5389]
**Do you want to request a *feature* or report a *bug*?**
bug

**What is the current behavior?**
When editing embedded document, validateSync causes an error.
It does not fail when editing a nested document and it does not fail when using validate() instead of validateSync()

**If the current behavior is a bug, please provide the steps to reproduce.**

This is my schema file:
```javascript
'use strict';

const mongoose = require('mongoose');
const dbConnection = require('./db-connection');

const ChildSchema = new mongoose.Schema(
    {
        _id: false,
        child_name: {
            type: String,
            required: true
        }
    }
);

const ParentSchema = new mongoose.Schema(
    {
        father_name: {
            type: String,
            required: true
        },
        child: ChildSchema,
        children: [ChildSchema]
    },
    {
        versionKey: false
    }
);

module.exports = dbConnection.model('family', ParentSchema, 'hierarchy');
```

let's assume that the starting point is this mongo document of a father with nested child and child array:

```javascript
{
    "_id" : ObjectId("594b58753270f9bcaba54b63"),
    "father_name" : "father",
    "child" : {
        "child_name" : "nested_child"
    },
    "children" : [ 
        {
            "child_name" : "array_child"
        }
    ]
}
```

if I return the doc and set the full `child` nested doc, validate and validateSync works perfectly fine:

```javascript
router.post(
    '/',
    function (req, res, next) {
        return hierarchyModel.findOne().exec().then(
            (resultHierarchy) => {
                resultHierarchy.child = {child_name: 'nested_child'};
                //resultHierarchy.children[0] = {child_name: 'array_child'};

                resultHierarchy.validateSync();

                return true;

                // return new Promise(function (resolve, reject) {
                //     resultHierarchy.validate((error) => {
                //         if (error) {
                //             return reject(error);
                //         } else {
                //             return resolve(resultHierarchy.save());
                //         }
                //     });
                // })
            }
        ).then(
            () => {
                return res.end();
            }
        ).catch(
            (errorInst) => {
                return res.status(500).end(errorInst.message);
            }
        )
    }
);
```

but if I edit the embedded child of the `children` array:

```javascript
router.post(
    '/',
    function (req, res, next) {
        return hierarchyModel.findOne().exec().then(
            (resultHierarchy) => {
                //resultHierarchy.child = {child_name: 'nested_child'};
                resultHierarchy.children[0] = {child_name: 'array_child'};

                resultHierarchy.validateSync();

                return true;

                // return new Promise(function (resolve, reject) {
                //     resultHierarchy.validate((error) => {
                //         if (error) {
                //             return reject(error);
                //         } else {
                //             return resolve(resultHierarchy.save());
                //         }
                //     });
                // })
            }
        ).then(
            () => {
                return res.end();
            }
        ).catch(
            (errorInst) => {
                return res.status(500).end(errorInst.message);
            }
        )
    }
);
```
causes the error:
**doc.validateSync is not a function**

but if I run it with the `validate` function, there is no error and the document is saved successfully.

**What is the expected behavior?**
validation capabilities should be consistent for embedded array and nested doc and for validate and validateSync functions.

**Please mention your node.js, mongoose and MongoDB version.**
node.js LTS 6.10.2
mongoose 4.10.8 and 4.7.1
MongoDB 3.4

**Comment by varunjayaraman** [comment]
thanks for the issue, will take a look soon

**Comment by varunjayaraman** [comment]
hm, I'm not able to reproduce your issue with a much simpler script:

```javascript
const mongoose = require('mongoose');
const co = require('co');
mongoose.Promise = global.Promise;
const GITHUB_ISSUE = `gh-5389`;


exec()
  .then(() => {
    console.log('successfully ran program');
    process.exit(0);
  })
  .catch(error => {
    console.error(`Error: ${ error }\n${ error.stack }`);
  });


function exec() {
  return co(function*() {
    const db = mongoose.createConnection(`mongodb://localhost:27017/${ GITHUB_ISSUE }`);

    const childSchema = new mongoose.Schema({
      _id: false,
      name: String,
      age: Number
    });

    const schema = new mongoose.Schema({
      name: String,
      children: [childSchema]
    });

    const Model = db.model('Model', schema);

    yield Model.remove({});
    
    const doc = yield Model.create({
      name: 'test',
      children: [
        { name: 'test-child', age: 24 }
      ]
    });

    doc.children[0].validateSync();
  });
}
```

**Comment by tomgrossman** [comment]
@varunjayaraman  that's because you didn't edit the child, you just created it.
before your last step ( `doc.children[0].validateSync();` ) and after you create it, you should `find` the `father` doc, then do the `doc.children[0] = { name: 'test-child', age: 24 }` and only then when you `validateAsync` it, you will get the error

**Comment by varunjayaraman** [comment]
ok i was able to reproduce, but i think what you're asking for is difficult without ES6 proxies, and can be solved by using `Document#set`:

```javascript
const mongoose = require('mongoose');
const co = require('co');
mongoose.Promise = global.Promise;
const GITHUB_ISSUE = `gh-5389`;


exec()
  .then(() => {
    console.log('successfully ran program');
    process.exit(0);
  })
  .catch(error => {
    console.error(`Error: ${ error }\n${ error.stack }`);
  });


function exec() {
  return co(function*() {
    const db = mongoose.createConnection(`mongodb://localhost:27017/${ GITHUB_ISSUE }`);

    const childSchema = new mongoose.Schema({
      _id: false,
      name: String,
      age: Number
    });

    const schema = new mongoose.Schema({
      name: String,
      children: [childSchema]
    });

    const Model = db.model('Model', schema);

    yield Model.remove({});
    
    const doc = yield Model.create({
      name: 'test',
      children: [
        { name: 'test-child', age: 24 }
      ]
    });

    const created = yield Model.findById(doc._id);

    // this doesnt beause it bypasses change detection
    created.children[0] = { name: 'updated-child', age: 53 };

    // this works
    // created.set('children.0', { name: 'updated-child', age: 53 });
    created.children[0].validateSync();
  });
}
```

**Comment by vkarpov15** [comment]
For posterity, this looks like essentially a duplicate of #3816 but for `validateSync()`.  Fix will be in 4.11.2
